File: JavaScript/294-Flip Game II.js
/**
 * @param {string} s
 * @return {boolean}
 */
var canWin = function(s) {
    let memo = new Map();
    
    let dfs = (s) =>{
        if (memo.has(s)) return memo.get(s);
        for (let i=0; i<s.length-1; i++){
            if (s[i] === '+' && s[i+1] === '+'){
                if (!dfs(s.substring(0,i) + '--' + s.substring(i+2))){
                    memo.set(s, true);
                    return true;
                }
            }
        }
        memo.set(s, false);
        return false;
    }
    return dfs(s);
};

File: 294-Flip Game II.py
class Solution(object):
    def canWin(self, s):
        """
        :type s: str
        :rtype: bool
        """
        memo = dict()
        
        def dfs(s):
            if s in memo:
                return memo[s]
            for i in range(1, len(s)):
                if s[i-1:i+1] == "++" and not dfs(s[:i-1] + "--" + s[i+1:]):
                    memo[s] = True
                    return True
            memo[s] = False
            return False
        
        return dfs(s)